README.md
# SmartCacheStore

SmartCacheStore provides variations on ActiveSupport's MemCacheStore and RedisCacheStore to make the default configurations more production ready. It does this by setting better defaults and using a few more ENVs.


##### Redis -- SmartRedisCache

Set by default:

    url:                ENV['REDIS_URL']
    namespace:          ENV['REDIS_NAMESPACE']
    expires_in:         1.day
    race_condition_ttl: 5.seconds
    pool:               ENV['RAILS_MAX_THREADS'] || 5
    connect_timeout:    2
    read_timeout:       1
    write_timeout:      1
    reconnect_attempts: [0.2]

`:url` should be a uri, like: `redis://localhost:6379/0`. For two or more uris, combine with a comma: `redis://host-a,redis://host-b`.

Any of these may still be specified directly, as well as all other support parameters for RedisCacheStore or the Redis gem.


##### Memcache -- SmartMemCache

Set by default:

    url:                ENV['MEMCACHE_SERVERS']
    namespace:          ENV['MEMCACHE_NAMESPACE']
    expires_in:         1.day
    race_condition_ttl: 5.seconds
    failover:           true
    pool:               ENV['RAILS_MAX_THREADS'] || 5

MemCacheStore's legacy initializer with `*addresses` as the first parameter, and the RedisCacheStore compatible `:url` are supported.

`:url` or *addresses should be a uri, like: `memcached://localhost:11211`. For two or more uris, combine with a comma: `memcached://host-a,memcached://host-b`.


### Other features

SmartMemCache's increment() and decrement() are also enhanced to be able to act on previously unset keys.


## Usage

In `environment/[env].rb`:

#### Redis

    # basic:
    config.cache_store = :smart_redis_cache
    # with parameters:
    config.cache_store = :smart_redis_cache, {url: 'redis://localhost:1234/9'}

#### Memcache

    # basic:
    config.cache_store = :smart_mem_cache
    # with parameters:
    config.cache_store = :smart_mem_cache, 'memcached://localhost:1234'
    config.cache_store = :smart_mem_cache, 'memcached://localhost:1234', {expires_in: 1.hour}
    config.cache_store = :smart_mem_cache, {url: 'memcached://localhost:1234', expires_in: 1.hour}



## Installation
Add `smart_cache_store` plus one driver to your application's Gemfile:

```ruby
gem 'redis' # for redis; gem v4.x or 5.x
# OR
gem 'dalli' # for memcache

gem "smart_cache_store"
```

And then execute:
```bash
$ bundle
```


## Contributing

1. Fork it ( https://github.com/zarqman/smart_cache_store/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

## License
The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).


CHANGELOG.md
#### 1.0.0

- (Breaking) Require ActiveSupport 6.1+ (drops 5.2, 6.0)

#### 0.3.1

- Change pool: N to pool: {size: N} for AS 7.1

#### 0.3.0

- Support ActiveSupport 7.1

#### 0.2.0

- Support redis-rb 5.x
- Add dependency on connection_pool

#### 0.1.0

- Initial release


test/smart_cache_store_test.rb
require "test_helper"

class SmartCacheStoreTest < ActiveSupport::TestCase
  test "it has a version number" do
    assert SmartCacheStore::VERSION
  end
end


test/dummy/config/application.rb
require_relative "boot"

require "rails"
# Pick the frameworks you want:
require "active_model/railtie"
# require "active_job/railtie"
# require "active_record/railtie"
# require "active_storage/engine"
require "action_controller/railtie"
# require "action_mailer/railtie"
# require "action_mailbox/engine"
# require "action_text/engine"
require "action_view/railtie"
# require "action_cable/engine"
# require "sprockets/railtie"
require "rails/test_unit/railtie"

# Require the gems listed in Gemfile, including any gems
# you've limited to :test, :development, or :production.
Bundler.require(*Rails.groups)
require "smart_cache_store"

module Dummy
  class Application < Rails::Application
    config.load_defaults Rails::VERSION::STRING.to_f

    # Configuration for the application, engines, and railties goes here.
    #
    # These settings can be overridden in specific environments using the files
    # in config/environments, which are processed later.
    #
    # config.time_zone = "Central Time (US & Canada)"
    # config.eager_load_paths << Rails.root.join("extras")
  end
end


test/dummy/config/environments/production.rb
require "active_support/core_ext/integer/time"

Rails.application.configure do
  # Settings specified here will take precedence over those in config/application.rb.

  # Code is not reloaded between requests.
  config.cache_classes = true

  # Eager load code on boot. This eager loads most of Rails and
  # your application in memory, allowing both threaded web servers
  # and those relying on copy on write to perform better.
  # Rake tasks automatically ignore this option for performance.
  config.eager_load = true

  # Full error reports are disabled and caching is turned on.
  config.consider_all_requests_local       = false
  config.action_controller.perform_caching = true

  # Ensures that a master key has been made available in either ENV["RAILS_MASTER_KEY"]
  # or in config/master.key. This key is used to decrypt credentials (and other encrypted files).
  # config.require_master_key = true

  # Disable serving static files from the `/public` folder by default since
  # Apache or NGINX already handles this.
  config.public_file_server.enabled = ENV["RAILS_SERVE_STATIC_FILES"].present?

  # Enable serving of images, stylesheets, and JavaScripts from an asset server.
  # config.asset_host = "http://assets.example.com"

  # Specifies the header that your server uses for sending files.
  # config.action_dispatch.x_sendfile_header = "X-Sendfile" # for Apache
  # config.action_dispatch.x_sendfile_header = "X-Accel-Redirect" # for NGINX

  # Force all access to the app over SSL, use Strict-Transport-Security, and use secure cookies.
  # config.force_ssl = true

  # Include generic and useful information about system operation, but avoid logging too much
  # information to avoid inadvertent exposure of personally identifiable information (PII).
  config.log_level = :info

  # Prepend all log lines with the following tags.
  config.log_tags = [ :request_id ]

  # Use a different cache store in production.
  # config.cache_store = :mem_cache_store


  # Enable locale fallbacks for I18n (makes lookups for any locale fall back to
  # the I18n.default_locale when a translation cannot be found).
  config.i18n.fallbacks = true

  # Don't log any deprecations.
  config.active_support.report_deprecations = false

  # Use default logging formatter so that PID and timestamp are not suppressed.
  config.log_formatter = ::Logger::Formatter.new

  # Use a different logger for distributed setups.
  # require "syslog/logger"
  # config.logger = ActiveSupport::TaggedLogging.new(Syslog::Logger.new "app-name")

  if ENV["RAILS_LOG_TO_STDOUT"].present?
    logger           = ActiveSupport::Logger.new(STDOUT)
    logger.formatter = config.log_formatter
    config.logger    = ActiveSupport::TaggedLogging.new(logger)
  end

  # Inserts middleware to perform automatic connection switching.
  # The `database_selector` hash is used to pass options to the DatabaseSelector
  # middleware. The `delay` is used to determine how long to wait after a write
  # to send a subsequent read to the primary.
  #
  # The `database_resolver` class is used by the middleware to determine which
  # database is appropriate to use based on the time delay.
  #
  # The `database_resolver_context` class is used by the middleware to set
  # timestamps for the last write to the primary. The resolver uses the context
  # class timestamps to determine how long to wait before reading from the
  # replica.
  #
  # By default Rails will store a last write timestamp in the session. The
  # DatabaseSelector middleware is designed as such you can define your own
  # strategy for connection switching and pass that into the middleware through
  # these configuration options.
  # config.active_record.database_selector = { delay: 2.seconds }
  # config.active_record.database_resolver = ActiveRecord::Middleware::DatabaseSelector::Resolver
  # config.active_record.database_resolver_context = ActiveRecord::Middleware::DatabaseSelector::Resolver::Session
end


lib/active_support/cache/smart_mem_cache.rb
# requires gems: dalli, connection_pool
#
# MEMCACHE_SERVERS can be either
#   server1:11211,server2
# or
#   memcached://user:pass@server1:11211,...
# default: localhost:11211


require 'active_support/cache/mem_cache_store'

module ActiveSupport
  module Cache
    class SmartMemCache < MemCacheStore

      # known options:
      #   ActiveSupport::Cache - universal options
      #     :namespace{nil}, :compress{true}, :compress_threshold{1k}, :coder, :expires_in{0}
      #     :race_condition_ttl{0}, :skip_nil{false}
      #   ActiveSupport::Cache::Store
      #     pool: {:size{5}, :timeout{5}}
      #   Dalli::Client
      #     :failover{true}, :serializer{Marshall}, :compressor{zlib}, :cache_nils{false}
      #     :namespace{nil}, :expires_in{0}, :compress{false} (separate from AS::C's same options above)
      #     :threadsafe{true} (using ConnPool turns this off automatically)
      def initialize(*addresses, **args)
        args.reverse_merge!(
          namespace:          ENV['MEMCACHE_NAMESPACE'],
          expires_in:         1.day,
          race_condition_ttl: 5.seconds,
          failover:           true,
        )
        if ActiveSupport.version >= '7.1.0.a'
          args[:pool] ||= {}
          args[:pool][:size] ||= Integer(ENV.fetch('RAILS_MAX_THREADS'){ 5 })
        else
          args[:pool_size] ||= Integer(ENV.fetch('RAILS_MAX_THREADS'){ 5 })
        end
        addresses.push Array(args.delete(:url)) if addresses.empty? && args.key?(:url)
        super(*addresses, args)
      end


      if ActiveSupport.version < '7.1.0.a'
        # MemCacheStore#increment docs say it will init any invalid value to 0.
        # In reality, it will only increment a pre-existing, raw value. everything else returns nil or an error.
        # This fixes init of missing value on both increment() and decrement().
        # Preexisting, invalid values still return errors.

        def increment(name, amount = 1, options = nil)
          options = merged_options(options)
          instrument(:increment, name, amount: amount) do
            rescue_error_with nil do
              @data.with { |c| c.incr(normalize_key(name, options), amount, options[:expires_in], amount) }
            end
          end
        end

        def decrement(name, amount = 1, options = nil)
          options = merged_options(options)
          instrument(:decrement, name, amount: amount) do
            rescue_error_with nil do
              @data.with { |c| c.decr(normalize_key(name, options), amount, options[:expires_in], 0) }
            end
          end
        end

      end

    end
  end
end


lib/active_support/cache/smart_redis_cache.rb
# requires gems: redis, connection_pool
#
# REDIS_URL (or :url) can have one or multiple urls:
#   redis://localhost:1234
#   redis://host-a,redis://host-b   # automatically uses Redis::Distributed


require 'active_support/cache/redis_cache_store'

module ActiveSupport
  module Cache
    class SmartRedisCache < RedisCacheStore

      # known options:
      #   ActiveSupport::Cache - universal options
      #     :namespace{nil}, :compress{true}, :compress_threshold{1k}, :coder, :expires_in{0}
      #     :race_condition_ttl{0}, :skip_nil{false}
      #   ActiveSupport::Cache::Store
      #     pool: {:size{5}, :timeout{5}}
      #   ActiveSupport::Cache::RedisCacheStore
      #     :error_handler
      #   Redis
      #     :host, :port, :db, :username, :password, :url, :path
      #     ssl_params: {:ca_file, :cert, :key}
      #     :sentinels, :role
      #     :replica
      #     :timeout (sets next 3), :connect_timeout, :read_timeout, :write_timeout
      #     :reconnect_attempts
      def initialize(**args)
        args.reverse_merge!(
          namespace:          ENV['REDIS_NAMESPACE'],
          expires_in:         1.day,
          race_condition_ttl: 5.seconds,
          connect_timeout:    2,
          read_timeout:       1,
          write_timeout:      1,
        )
        if ActiveSupport.version >= '7.1.0.a'
          args[:pool] ||= {}
          args[:pool][:size] ||= Integer(ENV.fetch('RAILS_MAX_THREADS'){ 5 })
        else
          args[:pool_size] ||= Integer(ENV.fetch('RAILS_MAX_THREADS'){ 5 })
        end
        if Redis::VERSION >= '5'
          args.reverse_merge!(
            # when an array, contains delay for each reconnect attempt
            reconnect_attempts: [0.2]
          )
        else
          args.reverse_merge!(
            reconnect_attempts: 1,
            reconnect_delay:    0.2,
          )
        end
        args[:url] ||= ENV['REDIS_URL'] unless args.key?(:redis) || args.key?(:url)
        args[:url] = args[:url].split(',') if args[:url].is_a?(String)
        super(**args)
      end

    end
  end
end
